### DISCUSSION TITLE
In-App Reporting using MeasureReports and support for a new configurable design

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#1735] [Ideas]
**Context** 
At the moment In-app reporting and MeasureReporting do not match one-to-one in implementation and reporting both on implementation and report view, thus we need to do some research and design on how to adopt this based on the specification and achieve the same level of accurate reporting to support the workflows for both CHAs and CHSS.

**Measure reporting interface** 
We need to figure out a way in which we can be able to update the previous In-App reporting to conform to Measurereporting based on the FHIR specification.

**Proposed Solution** 
At the moment we have implemented measure reporting by invoking the [$evalute_measure operation](http://fhir-ru.github.io/operation-measure-evaluate-measure.html) in the Android FHIR SDK to allow FHIR Core to be able to generate measure reports. We thus need to add a few components to make it compatible with the In-App reporting mechanism 

**Feature Scope** 
The scope of this work is limited to providing an interface to filter the desired subjects and periods for reporting, set the coverage for the indicators and then view aggregated reports, charts and graphs for reporting purposes. 

**In-App Reporting** 
The "In-App Reporting" feature empowers the health worker to see how she is performing against her daily task list and allows for offline peer-to-peer syncing of the data collected on a daily basis. This greatly simplifies reporting requirements by eliminating the need for cumbersome paper registries and tally sheets.

**Data collection and extraction (StructureMaps)**
1. We need to extract the appropriate resources for CQL queries to run and generate the indicators as MeasureReports 
2. We need to append the correct standardization codes (SNOMED/LOINC) to be able to run appropriate/simplified queries appropriately tagging the relevant resource 

**Tabulation of indicators** 
1. Types of MeasureReports  - This can only be determined once we have a description of the requirements for each indicator definition 
        - Proportion  - **In use**
        - Ratio - _Needs some RnD_
        - Continuous Variable  - _Needs some RnD_
        - Cohort - _Needs some RnD_
3. CQL authoring for measures/CQL Libraries 
        - Common Libraries  - **DONE**
        - Primary Logic Libraries    - **DONE**
        - Filter by the practitioner  - _Needs some RnD as this has not yet been done on the $evaluate_measure SDK API_

**Transformation of data for user interpretation**
1. How does the app consume generated MeasureReport FHIR Resources to the Data Visualizations (Charts and Graphs)?

**Reporting Interface**
1. New Reporting UI for eCBIS Indicators - Scoping and Implementation plan
        - Development of [Mock-ups ](https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=15615%3A103798)- DONE
        - Update of mock-ups based on inputs - TBD
        - Group work to Prepare Indicator Matrix for Dashboard Visualization - TBD

**Sync Data** 
We have the following mechanisms in place 
1. Configuration for the measure to be loaded into the app 
2. Configuration to have modules and related indicators 
4. Options generate  - Compiled measures at runtime or ~Stored Measures 

**Params for $evaluate_measure**
1. Date selection  - Generator for looping through year/months (Required)
2. Interface to pass through the 
4. Subject  = population
5. Practioner  - _This is not functional from the preliminary test of the $evlaute_measure API interface on Android FHIR SDK, filter may have to be done from CQL, implication on performance not known for now_ 

**Generation and bundling of Measure** 
For the purposes of this scoping the MeasureReports may be generated on-demand (for the current month) or as a background task (for the previous past months)
Afterwards, we need to see how these MeasureReports representing indicators can be :
1. Aggregated  per module  (Case in point using[ e-CBIS Indicators](https://docs.google.com/spreadsheets/d/1Kfp0rRYlksrBoSIecAuovZ8Y3k8yH9mkmgHcpQbMtXI/edit#gid=2066197482))
2. We may need a Data Source Interface to aggregate the measures based on modules to generate the charts/graph data 
3. Data visualisation configs - determines the Chart type, Legend/Keys, Color coding 

**Open Questions** 
1. What type of views to implement  - Types of charts and graphs etc. 
2. Key  management 
3. Creating charts via Compose 
6. Sample evaluation to generate sample data for consumption to the visualization component 
7. Compatible libraries for Kotlin and JetPack compose
        - https://semicolonspace.com/pie-chart-android-compose/
        - https://www.geeksforgeeks.org/pie-chart-in-android-using-jetpack-compose/ 

**Gaps/Assumptions** 
1. How do we Filter reports by Practitioner  - CHSS, CHW/CHA?
2. How do we add multiple data sets for the X and Y axis for a line graph, histogram, bar graph etc? 
3. How do we add targets for various indicators e.g Coverage

The implementation approach will involve  having an MVP and then an updated Alpha version 

**Next Steps** 
- [x] 1. Create issues for MVP 
- [ ] 2. Scope the Alpha version 

**Read more here**  
1. [Clinicalreasoning-quality-reporting - FHIR v4.6.0](https://build.fhir.org/clinicalreasoning-quality-reporting.html)  
2. [HL7.FHIR.US.CQFMEASURES\Glossary - FHIR v4.0.1](https://build.fhir.org/ig/HL7/cqf-measures/glossary.html) 
3. [https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html](https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html#ecqms)

**Comment by pld** [comment]
How are Measures linked to Compositions now?

**Reply by f-odhiambo** [reply]
We have it as a measure config in the [app configs ](https://github.com/opensrp/fhir-resources/tree/main/ecbis/app_configurations/ecbis_cha_preview). See here https://github.com/opensrp/fhir-resources/blob/main/ecbis/app_configurations/ecbis_cha_preview/measure_report_config.json

**Reply by pld** [reply]
OK cool, so we could structure sets of reports by changing the value for reports to a dictionary, like

```json
...
"reports": {
    "Household": [
		{
			"id": "1",
			"title": "Household Members",
			"description": "Number of Households, Household members registered, disaggregated by age and gender for each age group",
			"url": "http://fhir.org/guides/who/anc-cds/Measure/HOUSEHOLDIND01"
		},
		{
			"id": "2",
			"title": "Household Members ANC",
			"description": "Number of Households, Household members enrolled into ANC, disaggregated by age",
			"url": "http://fhir.org/guides/who/anc-cds/Measure/HOUSEHOLDPREGNANCY"
		}
    ],
    ...
}
```

the keys are the name of the set of reports, then when you click in you'd see the results for everything in the list of values.

**Reply by f-odhiambo** [reply]
> OK cool, so we could structure sets of reports by changing the value for reports to a dictionary, like
> 
> ```json
> ...
> "reports": {
>     "Household": [
> 		{
> 			"id": "1",
> 			"title": "Household Members",
> 			"description": "Number of Households, Household members registered, disaggregated by age and gender for each age group",
> 			"url": "http://fhir.org/guides/who/anc-cds/Measure/HOUSEHOLDIND01"
> 		},
> 		{
> 			"id": "2",
> 			"title": "Household Members ANC",
> 			"description": "Number of Households, Household members enrolled into ANC, disaggregated by age",
> 			"url": "http://fhir.org/guides/who/anc-cds/Measure/HOUSEHOLDPREGNANCY"
> 		}
>     ],
>     ...
> }
> ```
> 
> the keys are the name of the set of reports, then when you click in you'd see the results for everything in the list of values.

@pld 

From our discussion, with @ahsanbhatti49 and @ellykits  this will be in the below format 
```json
...
"reports": {
	"id": "1",
	"title": "Household Members",
	"description": "Number of Households, Household members registered, disaggregated by age and gender for each age group",
	"url": "http://fhir.org/guides/who/anc-cds/Measure/HOUSEHOLDIND01",
                       "module" : "Module 1 indicators"
        },
        {
	"id": "2",
	"title": "Household Members ANC",
	"description": "Number of Households, Household members enrolled into ANC, disaggregated by age",
	"url": "http://fhir.org/guides/who/anc-cds/Measure/HOUSEHOLDPREGNANCY",
                       "module" : "Module 1 indicators"
        }
```
}

**Comment by pld** [comment]
so you'd split the modules by grouping on the module key after parse? Why that versus nested? Seems like more computational effort to do the grouping, but maybe trivially more idk

**Reply by ellykits** [reply]
We compromised the computational for more flexibility in modeling the data structure and scalability. Using the module as the key would require a finite number of supported modules otherwise we would have to update the code each time we introduce a new reporting module.

**Reply by ellykits** [reply]
Actually I think we can still use the nested approach if we use a `Map<String, List<ReportConfig>`. We will still have the flexibility, **however** we will have problems with translations;  it is the `value` that needs to be translated not the `key`.

**Reply by ellykits** [reply]
The assumption above is the name of the report module is free text that should be displayed to the user and it needs to be translatable.

**Reply by ellykits** [reply]
To optimize the computation we can ensure the grouping only happens once.

**Reply by pld** [reply]
we could use an list of fhir list resources, they have title and entry keys in them, we'd make the title multilingual, https://www.hl7.org/fhir/list.html

**Reply by pld** [reply]
can we use a FHIR resource for the report config as well? Not sure what the appropriate resource would be, but the structure's pretty simple. I also don't remember why we have ID in there, maybe that was to give an ordering? We could have the field key be order if that's the only purpose, or find a FHIR resource with a field we can use for that.

If we need to discuss happy to do that, but we have more thinking to do before we make a decision here.

**Reply by ellykits** [reply]
I see. List resource uses references for the entries, meaning will need a reference of a Binary resource with all the relevant indicators for the module instead of a Binary resource per indicator.

**Reply by pld** [reply]
each entry in the list could be a separate indicator and the list itself is a group of indicators, then you have a binary (or other, there might be something more appropriate) resource per indicator

**Reply by ellykits** [reply]
Using FHIR resource/custom model entirely is okay. Combining custom models with FHIR resource adds the complexity of serialization.

**Reply by pld** [reply]
@dubdabasoduba pointed out we can have the entries in the list be the measures. So if we have a set of List resources, each List resource defines the module, then the entry points to the measure, that should satisfy our requirements here

**Comment by f-odhiambo** [comment]
New issues created 

**MVP**  
- [ ] 1. #1756
- [ ] 2. #1758
- [ ] 3. #1759
- [ ] 4. #1757
- [x] 5. #1885


**Alpha**  - To be scoped
- [ ] 1. #1867 
- [ ] 2. Incorporate the new UX with charts and graphs based on shared [mocks](https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=15615%3A103798)
